### PR TITLE
MSP OSD Revision

### PR DESCRIPTION
**Pull request by danielmohansahu**
## Overview

This pull request is an extension of #19515. From the original PR:

---
 This PR adds support for streaming telemetry over the MSP protocol. For now, it's specifically targeted to supporting the OSD functionality in the DJI Air Unit.

**msp_osd module functionality**
The new module implemented in this PR consumes data from uORB, and then immediately packs it into MSP messages to be transmitted over serial.

**What is MSP?**
MSP stands of MultiWii Serial Protocol. While the name has its roots in the outdated MultiWii project, MSP is used today in a variety of modern UAV applications for exchange of sensor data. The MSP protocol isn't concretely documented, but there are full implementation examples which can be found in Ardupilot, Betaflight, inav, etc. MSP is a simple serial based bytestream protocol. MSP exchanged can be RX only, TX only, or both request/response RXTX. For the purposes of this PR, we only TX.

--- 

## Summary of changes

This updated branch includes the following changes:
 * [x] Proper serial configuration via PX4 parameter
 * [x] Per-display icon toggling via PX4 parameter
 * [x] Visualization of platform state information and warning / error messages in the OSD.
 * [x] In-place refactoring to decouple `uorb` topic processing and string based visualization from, e.g., PX4 module boilerplate.

## Test data / coverage

Unit tests were added to verify the correctness of the `craft_name` character display.

## Potential improvements

A decent gain in processing efficiency is possible by figuring out which `uorb` topics to ignore when select displays are toggled off. This isn't quite as straightforward as it sounds, because the mapping from `uorb` topic -> display information isn't one-to-one (or many-to-one, or one-to-many). But if that mapping were deduced one could disable handling for whichever `uorb` topics aren't necessary for the given set of display icons.

### CODE REVIEW

File: boards/modalai/fc-v1/init/rc.board_defaults [hunk 1]
```diff
@@ -18,6 +18,12 @@
 # Common settings across Flight Core configurations
 #
 
+# Disable TELEM1 MAVLINK; used by OSD
+param set-default MAV_0_CONFIG 0
```
Comment (dagar) [inline boards/modalai/fc-v1/init/rc.board_defaults:22]: These aren't really defaults for the board itself.

**Comment by dagar** [comment]
I'll get https://github.com/PX4/PX4-Autopilot/pull/19515 cleaned up and merged so @chris1seto gets proper credit, then see if I can get this PR updated via rebase or merge without making too much of a mess.

**Comment by chris1seto** [comment]
Really nice job on this @danielmohansahu ! Have you tested this on real hardware yet? I plan to test this week and give it a proper review. For now, I really like the refactoring direction.

**Comment by danielmohansahu** [comment]
> I'll get #19515 cleaned up and merged so @chris1seto gets proper credit, then see if I can get this PR updated via rebase or merge without making too much of a mess.

Sounds good. For what it's worth I did my best to retain [the commit history](https://github.com/PX4/PX4-Autopilot/pull/20079/commits); my branch started off of @chris1seto . Hopefully that will make necessary modifications minimal.


> Really nice job on this @danielmohansahu ! Have you tested this on real hardware yet? I plan to test this week and give it a proper review. For now, I really like the refactoring direction.

Thanks! Yes, I've tested this on hardware. It performed as desired during testing. The warning message scrolling display is a little clunky, but it's probably as good as we can get for 14 characters. I did see one instance where the display cleared (no message, battery voltage showed 0.0, etc.) during a flight test, but I was unable to reproduce and chalked it up as a red herring.

**Comment by chris1seto** [comment]
> during a flight test, but I was unable to reproduce and chalked it up as a red herring.

The DJI osd does odd things sometimes so it wouldn't surprise me if this was their bug. Usually though this is in the form of a freeze or osd disappearing. I haven't yet seen values change up zero.

**Comment by dagar** [comment]
@danielmohansahu I merged @chris1seto's original work, then squashed your changes on top.

I wasn't able to push back to your branch. Can you test/review the squashed commit and hard reset your personal branch to finish this off? https://github.com/PX4/PX4-Autopilot/tree/feature/msp-osd-improvements-merge-candidate

**Comment by dagar** [comment]
Continued in https://github.com/PX4/PX4-Autopilot/pull/20297.

**Pull request closed without merging**